eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala

```diff
@@ -423,27 +422,47 @@ object Graph {
      * Representation of an edge of the graph
      *
      * @param desc        channel description
-     * @param update      channel info
+     * @param params      source of the channel parameters: can be a channel_update or hints from an invoice
      * @param capacity    channel capacity
      * @param balance_opt (optional) available balance that can be sent through this edge
      */
-    case class GraphEdge(desc: ChannelDesc, update: ChannelUpdate, capacity: Satoshi, balance_opt: Option[MilliSatoshi]) {
+    case class GraphEdge private(desc: ChannelDesc, params: ChannelRelayParams, capacity: Satoshi, balance_opt: Option[MilliSatoshi]) {
+
+      def this(u: ChannelUpdate, pc: PublicChannel) = this(
```

Comment:
We usually define constructors in the companion object (`apply`), what's the scala best practice here nowadays? I like the companion object approach, this way we only keep "business logic" inside the case class.

Reply:
Best practice is indeed to use `apply` methods in the companion object. Done in 8fcfebba3.

This gets rid of `new` keyword.